### PR TITLE
Blocks E2E: Refactor configs and workflow

### LINKED ISSUE #46331: Blocks E2E: Optimize sharding / CI resource usage
## Current Configuration Summary for the Blocks E2E Job

**Configurations**: 4  
**Total Shards**: 20 (5 per config)
**Total Tests**: 342 (as of April 30)

| Configuration              | Total Tests | Test Suites | Shards | Tests per Shard | Empty Shards |
|----------------------------|-------------|-------------|-------------------|-----------------|--------------|
| `block-theme` (default)    | 143         | 32          | 5                 | ~28            | 0            |
| `classic-theme`            | 11          | 5           | 5                 | ~2             |1          |
| `block-theme-with-templates` | 15          | 2           | 5                 | ~3               | 0          |
| `side-effects`             | 173         | 25          | 5                 | ~34            | 0            |

### Issues Identified

- **Resource Inefficiency**: The `classic-theme` and `block-theme-with-templates` configurations run very few tests per shard or even empty shards. 
- **Shard Imbalance**: The distribution of tests per shard is uneven, causing inefficient use of CI runners and potential delays in spinning up new runners (see pdV5qK-nT-p2).

### Step 1: Consolidate Test Groups

- **Combine Test Groups**: All tests should be grouped into a single configuration. This will simplify the management of test environments and enable optimal use of Playwright's automatic sharding feature.

### Step 2: Reduce Total Shards

- **Optimize Shards Number**: Reduce the total number of shards to 10. This configuration allows for approximately 34 tests per shard, optimizing the distribution based on the current test load.

### Expected Outcomes

- **Increased Efficiency**: Consolidating test groups and reducing the number of shards will improve the overall efficiency of resource usage.
- **Reduced Overhead**: Simplifying the configuration reduces complexity and management overhead.
- **Optimal Resource Utilization**: Leveraging automatic sharding ensures each CI runner is used to its full potential, avoiding underutilization and maximizing throughput.

This strategic adjustment aims to streamline the CI/CD pipeline for the Blocks E2E workflow, leading to better resource management and faster test execution times.

### PR DESCRIPTION
**Pull request by WunderBart**
Follow-up to https://github.com/woocommerce/woocommerce/pull/46125
Closes: #46331
Related: pdV5qK-nT-p2#possible-solutions

## Proposed Changes

This PR enhances the efficiency of running our Blocks E2E tests both locally and in our CI environment.

First, we have eliminated all `side effects` references as a continuation of #46125, ensuring they are permanently resolved. 🎉 

Regarding CI improvements, we've restructured the Playwright configuration (#46331) to enable all tests to run in a single job rather than a 3D theme-based matrix. This adjustment reduces resource usage by half (10 instead of 20 shards) and accelerates test completion due to better-balanced shards (cc: @tammullen). Additionally, we introduced a `Merge Artifacts` step, consolidating all debug artifacts from separate shards into one zip file for convenience.

For local testing improvements, we've reduced global setup rerun times by reusing existing database snapshots created within the initial run (introduced in #46125). We've also simplified the test execution process, making it unnecessary to identify the specific project (`block-theme`, `classic-theme`, or `block-theme-with-templates`) the test belongs to. Running a single suite is now as straightforward as using `pnpm test:e2e path-to/any-suite.spec.ts`. However, executing all theme-based groups remains an option through `pnpm test:e2e:block-theme/class-theme/block-theme-with-templates` when required.

## Testing Instructions

Ensure running tests via `pnpm test:e2e`, `pnpm test:e2e:block-theme`, `pnpm test:e2e:classic-theme`, `pnpm test:e2e:block-theme-with-templates` works as expected. Ensure all CI tests pass.

⚠️ Some tests might still be flaky. However, that flakiness is out of the scope of this PR, as it is no longer caused by the side effects. Those flaky tests need to be investigated individually.

## Next Steps

- Update [E2E docs](https://github.com/woocommerce/woocommerce/blob/8ee74ede0ebc1957089af08207ae2a1a8250610e/plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md).
- Explore uploading DB snapshot to the repo to reduce setup time.

### CODE REVIEW

File: plugins/woocommerce-blocks/tests/e2e/playwright-utils/test.ts [hunk 1]
```diff
@@ -161,12 +161,9 @@ const test = base.extend<
...
 		} );
 
-		const cliOutput = await cli(
+		await cli(
 			`npm run wp-env run tests-cli wp db import ${ DB_EXPORT_FILE }`
 		);
-		if ( ! cliOutput.stdout.includes( 'Success: Imported ' ) ) {
-			throw new Error( `Failed to import ${ DB_EXPORT_FILE }` );
-		}
 	},
```
Comment (Aljullu) [inline plugins/woocommerce-blocks/tests/e2e/playwright-utils/test.ts:167]: Out of curiosity: why this change?

**Review by Aljullu — APPROVED** [review]
Once again, thanks for all the improvements in our e2e suite, @WunderBart! :clap: 

I reviewed the code and tried some test commands, and everything looks good on my end. I left one question, but no need to change anything, I'm just curious. :slightly_smiling_face: 

I'm approving the PR, but given that it's a big one and I'm not an expert with e2e tests, I would appreciate if somebody else could take a second look just in case.

File: plugins/woocommerce-blocks/tests/e2e/playwright-utils/test.ts (hunk 1 above)
Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/playwright-utils/test.ts:167]: Nice catch! That check was needed because the `cli` utility always resolved the promise, regardless of whether the executed script threw an error. I made it reject on an error in https://github.com/woocommerce/woocommerce/pull/46409/commits/3a60fac18769ef218a0157ebf451f363ff6db73c, so now we don't need these explicit output checks as they're handled inside the utility.

File: .github/workflows/blocks-playwright.yml [hunk 2]
```diff
@@ -26,47 +25,44 @@ jobs:
...
+              uses: actions/upload-artifact@v4
+              if: ${{ !cancelled() }}
+              with:
+                  name: failures-artifacts-shard-${{ matrix.shardIndex }}
+                  path: plugins/woocommerce-blocks/tests/e2e/artifacts/test-results
+                  if-no-files-found: ignore
+    merge-artifacts:
+        name: Merge Artifacts
+        if: ${{ !cancelled() }}
+        needs: [blocks-playwright-tests]
```
Comment (gigitux) [inline .github/workflows/blocks-playwright.yml]: Great idea! ❤️

**Review by gigitux — APPROVED** [review]
Great work on the changes! LGTM!
One area for discussion:  Since the suffix approach is no longer required for environment setup, could we simplify the suffixes to "classic theme" and "block theme"? This would remove the redundant "block_theme_with_templates" suffix, but at the same time, make it easier to identify the theme during E2E tests: what do you think?

**Comment by WunderBart** [comment]
> Since the suffix approach is no longer required for environment setup, could we simplify the suffixes to "classic theme" and "block theme"? This would remove the redundant "block_theme_with_templates" suffix, but at the same time, make it easier to identify the theme during E2E tests: what do you think?

I need to figure out if I follow. IIUC, the suffixes are used to separate the themes we test against, which means we currently test against three different themes. What should we do with the suites now tagged with `block_theme_with_templates` if we remove the suffix?

**Comment by gigitux** [comment]
> > Since the suffix approach is no longer required for environment setup, could we simplify the suffixes to "classic theme" and "block theme"? This would remove the redundant "block_theme_with_templates" suffix, but at the same time, make it easier to identify the theme during E2E tests: what do you think?
> 
> I need to figure out if I follow. IIUC, the suffixes are used to separate the themes we test against, which means we currently test against three different themes. What should we do with the suites now tagged with `block_theme_with_templates` if we remove the suffix?

True, but other tests using a child theme starting from a classic theme and/or a block theme: https://github.com/woocommerce/woocommerce/blob/f339aee5b58e37e3d79229a176289a49fd925a63/plugins/woocommerce-blocks/tests/e2e/utils/constants.ts/#L6-L17


I would target them with the `.block_theme` because, in the end, `theme-with-woo-templates` is a block theme.

**Comment by WunderBart** [comment]
> > > Since the suffix approach is no longer required for environment setup, could we simplify the suffixes to "classic theme" and "block theme"? This would remove the redundant "block_theme_with_templates" suffix, but at the same time, make it easier to identify the theme during E2E tests: what do you think?
> > 
> > 
> > I need to figure out if I follow. IIUC, the suffixes are used to separate the themes we test against, which means we currently test against three different themes. What should we do with the suites now tagged with `block_theme_with_templates` if we remove the suffix?
> 
> True, but other tests using a child theme starting from a classic theme and/or a block theme: https://github.com/woocommerce/woocommerce/blob/f339aee5b58e37e3d79229a176289a49fd925a63/plugins/woocommerce-blocks/tests/e2e/utils/constants.ts/#L6-L17
> 
> I would target them with the `.block_theme` because, in the end, `theme-with-woo-templates` is a block theme.

Thanks for clarifying. Looking even further, do we even need the block/classic theme distinction? Do you think it's a useful thing to have? Previously, it was necessary to make the theme-based split since themes were being activated in the global setup phase. On this branch, we activate within the `beforeEach` hook, so the separation is unnecessary. I'm curious if anyone would actually be using the `pnpm test:e2e:classic-theme/block-theme` command since we usually don't run more than one suite locally, which can be done with `pnpm test:e2e path/to/suite.spec.ts`.

**Comment by gigitux** [comment]
> looking even further, do we even need the block/classic theme distinction? Do you think it's a useful thing to have?

Yeah, good point. I think that the classic/block suffixes are useful to have some separation. Otherwise, there is the risk that we will put everything in one file, and E2E tests will be huge.

**Comment by WunderBart** [comment]
> > looking even further, do we even need the block/classic theme distinction? Do you think it's a useful thing to have?
> 
> Yeah, good point. I think that the classic/block suffixes are useful for some separation. Otherwise, there is the risk that we will put everything in one file, and the E2E tests will be huge.

I also lean towards keeping them in separate files – it keeps the structure flat. I'll merge the `block-theme-with-templates` with `block-theme`, then! 

_edit:_

OTOH, there are only 11 tests for the classic theme, so I propose discussing the theme-based separation in another PR, if that's OK! 😅

**Comment by gigitux** [comment]
> OTOH, there are only 11 tests for the classic theme, so I propose discussing the theme-based separation in another PR, if that's OK! 😅

As I mentioned via DMs, my comment isn't a blocker 👍 On my side, this PR can be merged! :shipit:

**Comment by nielslange** [comment]
@WunderBart Firstly, thank you so much for working on this! This is another great improvement to the e2e tests. Regarding using different themes, what you and @gigitux discussed above, I recently created https://github.com/woocommerce/woocommerce/issues/46746. Not directly related to the issue raised by @gigitux, but a potential speed improvement when setting up the test env.
